### ISSUE TITLE
[Performance] Potential Performance optimzation for SDPA

### ISSUE THREAD

**Issue by drisspg** [pytorch/pytorch#100270]
# Summary
### Eager
The interface for SDPA is (batch, num_heads, seq_len, head_dim) to more closely align with the Attention is All You Need Paper. For most transformers architectures SDPA is called following an input Linear projection and this Projection is done on a tensor of shape (batch, seq_len, num_heads, head_dim).

For instance take a look at the CausalSelfAttention module used in the tutorial:
https://pytorch.org/tutorials/intermediate/scaled_dot_product_attention_tutorial.html#causal-self-attention.

This requries the query, key, value's to be transposed prior to calling SDPA. 

Both FlashAttention and MemEfficientAttention accept input q,k,v in the form: (batch, seq_len, num_heads, head_dim).  Within sdpa{flash, mem_eff} we call transpose here: 

For flash the following is the code pointer:
https://github.com/pytorch/pytorch/blob/c4bed869d16e9e0b761b82dda9130cfc14479141/aten/src/ATen/native/transformers/cuda/attention.cu#L715

In eager this manifests itself as two back-to-back view operations. This is not ideal but view ops are generally fairly cheap.

### PT2
Now we go to torch_compile which is great at removing this framework overhead. Since https://github.com/pytorch/pytorch/blob/c4bed869d16e9e0b761b82dda9130cfc14479141/aten/src/ATen/native/transformers/cuda/attention.cu#L688
is an aten operation that inductor doesn't know how to compile( this is good), inductor will fallback ot the aten op. However the transposes inside this ops which are potentially fusable can not be. 

### Question:
We could probably make sdpa_flash composite and just just at::_flash_forward as an aten op. This should in theory allow for the transposes to be compiled. Can we figure out the potential of this change before doing it? -> There is potential complexity with handiling NestedTensors but I think we could avoid the dispatcher and just do it manually.

**Comment by gchanan** [comment]
Could also run back memory_format (let's not).

**Comment by ngimel** [comment]
But in the common case the transposes are just view ops, so it doesn't matter whether inductor sees and compiles them or not?

**Comment by drisspg** [comment]
> We should make sure that inductor sends sdpa inputs in the format that sdpa wants (so that they become contiguous after sdpa internal reshape)

Yeah this in essence what I want to ensure is happening. And this issue was more of a book keeping task for me to investigate if that that is currently the case

**Comment by eellison** [comment]
If the internal of the operators dont involve any data movement (I think this is the case) - moving this outside of the kernel should be a negligible perf benefit.  there might be tiny overhead reduction

**Comment by drisspg** [comment]
Let me determine if different strides for dim 1 and 2 have different perf characteristics
